### PR TITLE
Make TestRunner wait for browser's idle when done (closes #162)

### LINKED ISSUE #162: Remote browser stops on the last page of the test run once task is finished
If you use remote browser for the testrun via `testcafe remote some-test.js` then test is completed remote just frozes on the last page of the test run. I suppose this happens because we terminate process before remote goes to the idle page. However, I think we should wait till remote rich the idle page and shown disconnection status there.

### PR DESCRIPTION
**Pull request by AndreyBelym**
\cc @inikulin @AlexanderMoskovkin

### CODE REVIEW

File: src/runner/index.js [hunk 1]
```diff
@@ -33,17 +32,15 @@ export default class Runner extends EventEmitter {
...
     }
 
     static async _freeBrowserConnection (bc, errorHandler) {
         bc.removeListener('error', errorHandler);
 
-        // NOTE: we should close local connections and
-        // related browsers once we've done
-        if (bc instanceof LocalBrowserConnection)
-            await Runner._waitForLocalBrowserClose(bc);
+        await Runner._waitForBrowserDisconnected(bc);
```
Comment (inikulin) [inline src/runner/index.js]: We can't close non-local connections, since they maybe shared with the other runner.

**Comment by inikulin** [comment]
\r-

**Comment by AndreyBelym** [comment]
FPR

**Comment by AndreyBelym** [comment]
FPR

**Comment by inikulin** [comment]
We've reached the point, when I can't understand how our browser connection management works :cry: We need to simplify things somehow, let's discuss it tomorrow.

**Comment by AlexanderMoskovkin** [comment]
The main point - we should wait for the idle page (and close the browser only for a local browser connection) before finish run task. I propose the following (it's the schematic example):
```js
//BrowserConnection
class BrowserConnection {
    waitForIdlePage () {
        this.waitForIdle = true;
        return Promise (function (resolve) {
            this.once('idle', resolve);
            setTimeout(resolve, 10000);
        });
    }

    close () {
        //real bc closing
        this.disconnected = true;
        this.ready        = false;

        this.browserConnectionGateway.stopServingConnection(this);
        clearTimeout(this.heartbeatTimeout);
        this.emit('close');
    }

    getStatus () {
        if(this.waitForIdle) {
            if(this.switchedToIdle) {
                this.waitForIdle = false;
                this.switchedToIdle = false; 
                this.emit('idle');
            }
            else {
                 this.switchedToIdle = true;
                 return { cmd: COMMAND.idle, url: this.idleUrl };
            }
        }
        ...
    }
}

//LocalBrowserConnection
class LocalBrowserConnection {
    close () {
        testCafeBrowserNatives.closeBrowser().then(()=>super.close());
    }
}

//Usage
runTests () {
    return new Promise (resolve => {
        ...
        task.on('done', async function () {
            await bc.waitForIdlePage();
            
            if(isLocalBrowserConnection) {
                bc.close();
                bc.once('close', resolve);
            }
            else
                resolve();
        });
    });
}
```

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

File: src/browser-connection/index.js [hunk 2]
```diff
@@ -70,11 +73,24 @@ export default class BrowserConnection extends EventEmitter {
...
         clearTimeout(this.heartbeatTimeout);
+
+        this.closed = true;
+        this.emit('closed');
+    }
+
+    switchToIdle () {
+        this.jobQueue = [];
+
+        this.waitingForIdleTimeout = setTimeout(() => {
```
Comment (AlexanderMoskovkin) [inline src/browser-connection/index.js]: `setTimeout` returns an `object`, so the default value for `this.waitingForIdleTimeout` is `null`

File: src/browser-connection/local.js [hunk 3]
```diff
@@ -33,31 +25,11 @@ export default class LocalBrowserConnection extends BrowserConnection {
     close () {
         // NOTE: When using local connections, we should close the browser before closing
         // the connection. For this, we close connections in the getStatus function.
```
Comment (AlexanderMoskovkin) [inline src/browser-connection/local.js]: the comment is obsolete

File: src/client/browser/index.js [hunk 4]
```diff
@@ -46,7 +46,7 @@ export function checkStatus (statusUrl, XMLHttpRequest) {
                 document.location = res.url;
 
             //NOTE: prepare to close the browser
-            if (res.cmd === COMMAND.close)
+            if (res.cmd === COMMAND.idle)
```
Comment (AlexanderMoskovkin) [inline src/client/browser/index.js]: set the document.title when page is loaded. `COMMAND.idle` is checked in the previous `if` statement

Comment (AlexanderMoskovkin) [inline src/client/browser/index.js]: we should do it in the `idle` page script

File: src/browser-connection/local.js [hunk 5]
```diff
@@ -1,19 +1,11 @@
 import { open as openBrowser, close as closeBrowser } from 'testcafe-browser-natives';
 import BrowserConnection from './index';
-import COMMAND from './command';
```
Comment (AlexanderMoskovkin) [inline src/browser-connection/local.js]: check the `browser-connection\command.js` file. The `COMMAND.close` command is not used yet

**Comment by AndreyBelym** [comment]
FPR

File: src/browser-connection/index.js [hunk 6]
```diff
@@ -91,6 +107,15 @@ export default class BrowserConnection extends EventEmitter {
     }
 
     renderIdlePage () {
+        if (this.waitingForIdleTimeout) {
+            clearTimeout(this.waitingForIdleTimeout);
```
Comment (inikulin) [inline src/browser-connection/index.js]: We can do it without condition

**Comment by AndreyBelym** [comment]
Closing the PR because #162 was merged in #131. I'll create a new PR to fix #131.

**Pull request closed without merging**